src/lib/core/crash.c

```diff
@@ -508,20 +508,28 @@ crash_report_stderr(struct crash_info *cinfo)
 static void
 crash_signal_cb(int signo, siginfo_t *siginfo, void *context)
 {
-	static volatile sig_atomic_t in_cb = 0;
-	struct crash_info *cinfo;
-
-	if (in_cb == 0) {
-		in_cb = 1;
-		cinfo = crash_collect(signo, siginfo, context);
-		crash_report_stderr(cinfo);
-		if (send_crashinfo &&
-		    crash_report_feedback_daemon(cinfo) != 0) {
-			pr_crit("unable to send a crash report");
+	/**
+	 * Some users of tarantool spawn threads and do not
+	 * block signals on them, so let's check that callback
+	 * is being executed on the main thread. If it is not -
+	 * we will only try to dump a core.
+	 */
```

Comment:
SIGSEGV and SIGFPE are thread directed signals - AFAIK they are always received by the thread that triggered the fault, even if they are blocked. AFAICS this patch basically breaks stack trace reporting for faults in threads other than tx (iproto, wal, etc). I think that `crash_collect` should be safe to call from any thread.

Reply:
> I think that crash_collect should be safe to call from any thread.
From any thread with initialized `cord` (at least, it uses `fiber()` macro for backtraces), so I changed condition `cord_is_main()` to `(cord() != NULL)`.

Reply:
AFAICS `backtace_collect()` doesn't use `fiber` if it equals self. I think that we should have `backtrace_collect_fiber(bt, fiber)` and `backtrace_collect(bt)`. The latter shouldn't depend on `fiber()` or `cord()`.

Anyway, the comment to the commit is stale.

Reply:
https://github.com/tarantool/tarantool/blob/f21f8e903a53b7c9ed11b689d4f8a10f50639003/src/lib/core/crash.c#L215-L220
`crash_collect()` uses `fiber()` as an argument for `backtace_collect()`. If `cord` is not initialized, `fiber()` macro will dereference a NULL pointer.

Reply:
> Anyway, the comment to the commit is stale.

What an inattention! Fixed it.

Reply:
`backtrace_collect()` implementation doesn't use the fiber if it equals `fiber()`.

Reply:
Dereferencing `NULL` is UB. Even if it is not used later.

Reply:
Please read again what I wrote [here](https://github.com/tarantool/tarantool/pull/7431#discussion_r927569882) and think.

Reply:
Sorry, I somehow missed that part. What if we just allow to pass `NULL` as fiber argument, and guarantee that module `fiber` won't be used in `backtrace_collect` if `NULL` is passed? (That's what I've done now).

Reply:
Sounds good to me.